Sources/Articles/Documentation.docc/2024/beginners-guide-to-protocol-buffers-and-grpc-with-swift/beginners-guide-to-protocol-buffers-and-grpc-with-swift.md

```diff
@@ -0,0 +1,360 @@
+import Logging
+import GRPC
+import NIO
+import GRPCCore
+
+@main
+struct Entrypoint: ParsableCommand {
+    
+    @Option(name: .shortAndLong)
+    var hostname: String = "127.0.0.1"
+    
+    @Option(name: .shortAndLong)
+    var port: Int = 1234
+    
+    func run() throws {
+        
+        var logger = Logger(label: "grpc-example")
+        logger.logLevel = .debug
+        
+        let group: EventLoopGroup = .singletonMultiThreadedEventLoopGroup
+
+        // 1. 
+        let server = Server.insecure(group: group)
+            .withServiceProviders(
+                [
+                    // 2.
+                    TodoService(),
+                ]
+            )
+            // 3.
+            .bind(to: .host(hostname, port: port))
+        
+        // 4.
+        server.map {
+            $0.channel.localAddress
+        }.whenSuccess { address in
+            logger.debug("gRPC Server started on port \(address!.port!)")
+        }
+        
+        // 5.
+        _ = try server.flatMap { $0.onClose }.wait()
+    }
+}
+```
+
+1.	The `Server.insecure(group:)` creates a gRPC server using the specified `EventLoopGroup` for handling requests.
+2.	The `withServiceProviders` method registers the `TodoService` instance, which contains the service logic for handling gRPC requests.
+3.	The `.bind(to:)` method binds the server to the specified hostname and port, making it ready to accept incoming requests.
+4.	The `server.map` retrieves the local address of the server and logs a message indicating that the gRPC server has successfully started on the specified port.
+5.	The application waits for the server to close before shutting down, keeping the server running to handle requests.
+
+To test the gRPC server, the [Evans gRPC client](https://github.com/ktr0731/evans) is a great option. Evans is a user-friendly, interactive gRPC client that allows developers to communicate with gRPC servers without needing to write any additional code. It provides a REPL-like interface for interacting with the server’s methods.
+
+Evans is available via the Homebrew package manager, making installation straightforward. Simply run the following command:
+
+```sh
+brew tap ktr0731/evans && brew install evans
+```
+
+Once installed, you can connect to your gRPC server by specifying the host and port:
+
+```sh
+evans repl --host localhost --port 1234 --proto ./todo.proto
+```
+
+This will launch Evans in interactive mode, allowing you to call the available RPC methods defined in your .proto files. Evans simplifies testing gRPC servers, making it a valuable tool during development and debugging.
+
+## Summary
+
+That’s all it takes to set up a basic gRPC server using Swift and Protocol Buffers. Many advanced features are available, such as streaming, bidirectional data flow, and generating client-side code. I hope this tutorial has provided a solid foundation for getting started with gRPC servers in Swift using the Swift Protobuf and gRPC libraries.
```

Comment:
Note: The tutorial currently doesn't link to the protobuf language guide. While I suggested one link up above, we should add more resources down here

Reply:
Could you recommend some links that you'd like to include? 🤔

Reply:
Definitely this one: https://protobuf.dev/programming-guides